src/coap_tinydtls.c

```diff
@@ -562,11 +698,454 @@ coap_get_tls_library_version(void) {
+  struct stat statbuf;
+
+  *length = 0;
+  if (!f)
+    return NULL;
+
+  if (fstat(fileno(f), &statbuf) == -1) {
+    fclose(f);
+    return NULL;
+  }
+
+  buf = coap_malloc(statbuf.st_size+1);
+  if (!buf)
+    return NULL;
+
+  if (fread(buf, 1, statbuf.st_size, f) != (size_t)statbuf.st_size) {
+    fclose(f);
+    free(buf);
+    return NULL;
+  }
+  buf[statbuf.st_size] = '\000';
+  *length = (size_t)(statbuf.st_size + 1);
+  fclose(f);
+  return buf;
+}
+
+static coap_binary_t *
+pem_decode_file_asn1(const char *begstr, const char *file)
+{
+  size_t length;
+  uint8_t *str = read_file_mem(file, &length);
+  char *bcp;
+  char *tcp;
+  coap_binary_t *decoded = NULL;
+
+  if (!str) {
+    return NULL;
+  }
+
+  bcp = strstr((const char*)str, begstr);
+  tcp = bcp ? strstr(bcp, "-----END ") : NULL;
+  if (bcp && tcp) {
+    bcp += strlen(begstr);
+    decoded = pem_base64_decode ((const uint8_t *)bcp, tcp - bcp);
+  }
+  coap_free(str);
+  return decoded;
+}
+#endif /* DTLS_ECC */
+
 int
-coap_dtls_context_set_pki(coap_context_t *ctx UNUSED,
-  coap_dtls_pki_t* setup_data UNUSED,
+coap_dtls_context_set_pki(coap_context_t *ctx,
+  coap_dtls_pki_t* setup_data,
   coap_dtls_role_t role UNUSED
 ) {
+#ifdef DTLS_ECC
+  coap_tiny_context_t *t_context;
+  coap_binary_t *asn1_priv;
+  coap_binary_t *asn1_pub;
+
+  if (!setup_data)
+    return 0;
+  if (setup_data->version != COAP_DTLS_PKI_SETUP_VERSION)
+    return 0;
+  if (!setup_data->is_rpk_not_cert)
+    return 0;
+  if (!ctx)
+    return 0;
+  t_context = (coap_tiny_context_t *)ctx->dtls_context;
+  if (!t_context)
+    return 0;
+  if (t_context->priv_key) {
+    coap_delete_binary(t_context->priv_key);
+    t_context->priv_key = NULL;
+  }
+  if (t_context->pub_key) {
+    coap_delete_binary(t_context->pub_key);
+    t_context->pub_key = NULL;
+  }
+  t_context->setup_data = *setup_data;
+
+  /* All should be RPK only now */
+  switch (setup_data->pki_key.key_type) {
+  case COAP_PKI_KEY_PEM:
+    if (setup_data->pki_key.key.pem.public_cert &&
+        setup_data->pki_key.key.pem.public_cert[0] &&
+        setup_data->pki_key.key.pem.private_key &&
+        setup_data->pki_key.key.pem.private_key[0]) {
+      /* Need to read in the PEM information and convert to binary */
+      asn1_priv = pem_decode_file_asn1("-----BEGIN EC PRIVATE KEY-----",
+                                 setup_data->pki_key.key.pem.private_key);
+      if (!asn1_priv) {
+        asn1_priv = pem_decode_file_asn1("-----BEGIN PRIVATE KEY-----",
+                                 setup_data->pki_key.key.pem.private_key);
+        if (!asn1_priv) {
+          coap_log(LOG_INFO, "Private Key (RPK) invalid\n");
+          return 0;
+        }
+      }
+      asn1_pub = pem_decode_file_asn1(
+                                "-----BEGIN PUBLIC KEY-----",
+                                setup_data->pki_key.key.pem.private_key);
+      if (!asn1_pub) {
+        coap_log(LOG_INFO, "EC Public Key (RPK) invalid\n");
```

Comment:
```
if (!asn1_pub) {
    asn1_pub = pem_decode_file_asn1("-----BEGIN EC PRIVATE KEY-----",
	                         setup_data->pki_key.key.pem.private_key);
    if (!asn1_pub) {
       coap_log(LOG_INFO, "EC Public Key (RPK) invalid\n");
       coap_delete_binary(asn1_priv);
       return 0;
    }
}
```

With that, it's possible to use the openssl output of 

> openssl ecparam -genkey -name prime256v1 -noout -out ec_pair.pem

directly without manipulations.

Reply:
Yes, in principle that works.  I need to do something similar in GnuTLS and update documentation before I push a fix - which should be in the next couple of hours or so.

Reply:
TinyDTLS code changes made - GnuTLS library expects both Private Key and Public Key, so looking at how to best abstract the Public Key and provide that separately.

Reply:
All code changes pushed including change for GnuTLS.